### DISCUSSION TITLE
Use GitHub Actions with a private runner to deploy to IIS

### DISCUSSION THREAD

**Discussion by giscus** [rajbos/rajbos.github.io#11] [Comments]
# Use GitHub Actions with a private runner to deploy to IIS

Use GitHub Actions with a private runner to deploy to IIS

https://devopsjournal.io/blog/2020/11/24/github-actions-with-private-runner-iis

**Comment by miyagisan221b** [comment]
Awesome article. Any ideas how to overcome "Could not create SSL/TLS secure channel" errors when it attempts to install build on-prem?

**Reply by rajbos** [reply]
Thanks!

I usually see this happen when we are installing the runners on a GHES instance that has a self signed certificate. That brings all sorts of errors with it 😕

**Reply by miragefoto** [reply]
Obviously this is an older post, but is stopping iis the best way to go?  IF there are any other sites running it drops them all, and then has to rebuild everything as it spins back up.  We were killing 20+ sites with every push to any of them.  We updated our deployment section to use  
- name: Deploy to IIS
      run: |
        # stop-webapppool {iis apppool name}
        # stop-iissite -Name {iis website name} -Confirm: $false
        # Copy-Item ./{same "publish" name from above}/* F:/Websites/{folder path} -Recurse -Force
        # start-webapppool {iis apppool name}
        # start-iissite {iis website name}

This keeps any other sites in IIS running.

**Reply by rajbos** [reply]
Good point indeed, that is a more elegant example indeed. This post was only focusing on how to do this from a self-hosted runner into an IIS server and the overall workings. Stopping the entire webserver is a bit much then 😉 . 

Thanks for sending in this improvement!

**Reply by invontodev** [reply]
Im looking to deploy to IIS on another machine, what would be the best way to achieve this?

**Reply by rajbos** [reply]
Use PowerShell remoting to connect to the other machine, then execute the commands in that session. 

More info here: https://docs.microsoft.com/en-us/powershell/scripting/learn/ps101/08-powershell-remoting

**Comment by oos-dev** [comment]
Muchas gracias por la ayuda, muy basica y simple.
Sin tantas complicaciones, ya pude configurar mi pipeline con mi servidor .

Algunos problemillas para copiar archivos pero al menos quedo listo.

**Reply by miragefoto** [reply]
We have updated our code.   We created an offline.html page which is stored on the server.  We then copy the offline.html file to the website folder, thus taking the site online, while displaying a placeholder message from the offline.html.  Then copy in the new files and remove the ffline.html will restart the site.  Sample code below.

`name: .NET Core Deploy to IIS
on:
  push:
    branches: 
      - master
      - main      

jobs:
  build-and-deploy:
    runs-on: self-hosted

    env: 
    # set apppool and site name from IIS
      AppPoolName : DefaultAppPool
      SiteName  : 'Default Web Site'
      SitePublishName  :  'DefaultWebSite'
    # set to site files.  In this case, the part of the path after f:/websites/
      SitePath :  MainWebsite
    steps:
    - uses: actions/checkout@v3

    - name: Build
      run: dotnet build --configuration Release .
    - name: dotnet publish 
      run: | 
        dotnet publish -c Release -o ${{env.SitePublishName}}Publish .
     
    - name: Look for app_offline add as needed
      run: |
        try{Get-Item F:/Websites/${{env.SitePath}}/app_offline.htm}
        catch{
          Copy-Item F:/Websites/Offline/app_offline.htm F:/Websites/${{env.SitePath}} -Recurse -Force  
          }
        
    - name: Sleep for 5 seconds
      run: Start-Sleep -s 5 
    
    - name: Copy Files
      run: |
        $currentRetry = 1;
        $backupRetry = 1;
        $success = $false;
        $backupSuccess = $false;
        do{
          echo "Attempting file copy $currentRetry"  
          try {
            Copy-Item ./${{env.SitePublishName}}Publish/* 'F:/Websites/${{env.SitePath}}' -Recurse -Force
            $success = $true;
          }
          catch {
            echo "Copy Failed due to following Catch error:`n"
            echo $PSItem.Exception.Message
            echo "`n"
            Start-Sleep -s 2
            $currentRetry = $currentRetry + 1;
          }
          finally {
            if ($currentRetry -ge 10) {
                echo "Copy keeps failing; restarting AppPool/Site as last resort`n"
                echo "Attempting to restart AppPool`n"
                do{
                    $status = Get-WebAppPoolState -name '${{env.AppPoolName}}'
                    if ($status.Value -eq "Stopped") {
                        start-WebAppPool ${{env.AppPoolName}}
                        echo "AppPool restarted`n---------`n"
                        $backupSuccess = $true;
                    }
                    else {
                        if ($backupRetry -ge 10) {
                           throw "AppPool restart keeps failing. Please look into Malfurion"
                        }
                        echo "AppPool not stopped yet; Re-attempt #$backupRetry"
                        Start-Sleep -s 3
                        $backupRetry = $backupRetry + 1;
                    }
                }
                while (!$backupSuccess -and $backupRetry -le 10)
                $backupRetry = 1;
                $backupSuccess = $false;
                echo "Attempting to restart Website`n"
                do{
                    $status = Get-WebsiteState -name '${{env.SiteName}}'
                    if ($status.Value -eq "Stopped") {
                        start-iissite ${{env.SiteName}}
                        echo "Website restarted`n---------`n"
                        $backupSuccess = $true;
                    }
                    else {
                        if ($backupRetry -ge 10) {
                           throw "Website restart keeps failing. Please look into Server"
                        }
                        echo "Website not stopped yet; Re-attempt #$backupRetry"
                        Start-Sleep -s 3
                        $backupRetry = $backupRetry + 1;
                    }
                }
                while (!$backupSuccess -and $backupRetry -le 10)
            }
          }
        }
        while (!$success -and $currentRetry -le 10) 
      
    - name: Remove Offline
      if: always()
      run: |
        Remove-Item F:/Websites/${{env.SitePath}}/app_offline.htm -Force
        
   
  notify:
    name: Discord Notification
    runs-on: ubuntu-latest
     # make sure the notification is sent AFTER the jobs you want included have completed
    needs:
      - build-and-deploy
    if: ${{ always() }} # You always want to be notified: success, failure, or cancelled

    steps:
      - name: Notify
        uses: nobrayner/discord-webhook@v1
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          discord-webhook: ${{ secrets.DISCORD_WEBHOOK }}
          username: 'GitHub Notification'
          avatar-url: 'https://octodex.github.com/images/Terracottocat_Single.png'
          title: '${{ github.repository }} publish: {{STATUS}}'
          description: '${{ github.event_name }} triggered this {{STATUS}}!'
          include-details: 'false'
          color-success: '#4287f5'
          color-failure: 'eb4034'
          color-cancelled: '0x42daf5'
`
